cmd/binlog-collector/collector/collector.go

```diff
@@ -0,0 +1,205 @@
+package collector
+
+import (
+	"bytes"
+	"log"
+	"os/exec"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/percona/percona-xtradb-cluster-operator/cmd/binlog-collector/db"
+	"github.com/percona/percona-xtradb-cluster-operator/cmd/binlog-collector/storage"
+)
+
+type Controller struct {
+	db              *db.PXC
+	storage         storage.Client
+	lastSet         string // last uploaded binary logs set
+	pxcServiceName  string // k8s service name for PXC, its for get correct host for connection
+	pxcUser         string // user for connection to PXC
+	pxcPass         string // password for connection to PXC
+	lastSetFileName string // name for object where the last binlog set will stored
+}
+
+type Config struct {
+	PXCServiceName string
+	PXCUser        string
+	PXCPass        string
+	S3Endpoint     string
+	S3AccessKeyID  string
+	S3AccessKey    string
+	S3BucketName   string
+}
+
+func New(c Config) (*Controller, error) {
+	s3, err := storage.NewS3(c.S3Endpoint, c.S3AccessKeyID, c.S3AccessKey, c.S3BucketName, true)
+	if err != nil {
+		return nil, errors.Wrap(err, "new storage manager")
+	}
+	lastSetFileName := "last-set"
+	// get last binlog set stored on S3
+	lastSet, err := s3.GetObject(lastSetFileName)
+	if err != nil {
+		return nil, errors.Wrap(err, "get lastset content")
+	}
+
+	return &Controller{
+		storage:         s3,
+		lastSet:         string(lastSet),
+		pxcUser:         c.PXCUser,
+		pxcPass:         c.PXCPass,
+		pxcServiceName:  c.PXCServiceName,
+		lastSetFileName: lastSetFileName,
+	}, nil
+}
+
+func (c *Controller) Run() error {
+	err := c.newDB()
+	if err != nil {
+		return errors.Wrap(err, "new db connection")
+	}
+	defer c.closeDB()
+
+	err = c.CollectBinLogs()
+	if err != nil {
+		return errors.Wrap(err, "collect binlog files:")
+	}
+
+	return nil
+}
+
+func (c *Controller) newDB() error {
+	host, err := c.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+	pxc, err := db.NewPXC(host, c.pxcUser, c.pxcPass)
+	if err != nil {
+		return errors.Wrapf(err, "new manager with host %s", host)
+
+	}
+	c.db = pxc
+
+	return nil
+}
+
+func (c *Controller) closeDB() error {
+	return c.db.Close()
+}
+
+func (c *Controller) getHost() (string, error) {
+	cmd := exec.Command("peer-list", "-on-start=/usr/bin/get-pxc-state", "-service="+c.pxcServiceName)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", errors.Wrap(err, "get output")
+	}
+	nodes := strings.Split(string(out), "node:")
+	sort.Strings(nodes)
+	lastHost := ""
+	for _, node := range nodes {
+		if strings.Contains(node, "wsrep_ready:ON:wsrep_connected:ON:wsrep_local_state_comment:Synced:wsrep_cluster_status:Primary") {
+			nodeArr := strings.Split(node, ".")
+			lastHost = nodeArr[0]
+		}
+	}
+	if len(lastHost) == 0 {
+		return "", errors.New("cant find host")
+	}
+
+	return lastHost + "." + c.pxcServiceName, nil
```

Comment:
nice to see how we cut the service name in line 104 and add it in line 112.
to avoid it it is possible to cut by `:` symbol, like
```
- nodeArr := strings.Split(node, ".")
+ nodeArr := strings.Split(node, ":")
```

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/622/commits/aabec733bdf055557998a691d5e502f99c0e9b1f